protobuf/coreapi/client_safe.go

```diff
@@ -0,0 +1,344 @@
+package coreapi
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/mesg-foundation/core/protobuf/acknowledgement"
+	"google.golang.org/grpc"
+)
+
+// ReconnectStrategy for grpc stream.
+type ReconnectStrategy int
+
+// Possible Reconnect Strategy for grpc stream.
+const (
+	Never ReconnectStrategy = iota
+	OnError
+)
+
+const reconnectDelay = 3 * time.Second
+
+// CoreClientSafe provides CoreClient with stream reconnection.
+type CoreClientSafe struct {
+	reconnect ReconnectStrategy
+	CoreClient
+}
+
+// NewCoreClient creates core client with reconnection.
+func NewCoreClientSafe(cc *grpc.ClientConn, rc ReconnectStrategy) *CoreClientSafe {
+	return &CoreClientSafe{
+		reconnect:  rc,
+		CoreClient: NewCoreClient(cc),
+	}
+}
+
+// listenEvent returns listen event client.
+func (s *CoreClientSafe) listenEvent(ctx context.Context, in *ListenEventRequest, opts ...grpc.CallOption) (Core_ListenEventClient, error) {
+	client, err := s.CoreClient.ListenEvent(ctx, in, opts...)
+	if err != nil {
+		return nil, err
+	}
+	if err := acknowledgement.WaitForStreamToBeReady(client); err != nil {
+		return nil, err
+	}
+	return client, err
+}
+
+// listenResult returns listen result client.
+func (s *CoreClientSafe) listenResult(ctx context.Context, in *ListenResultRequest, opts ...grpc.CallOption) (Core_ListenResultClient, error) {
+	client, err := s.CoreClient.ListenResult(ctx, in, opts...)
+	if err != nil {
+		return nil, err
+	}
+	if err := acknowledgement.WaitForStreamToBeReady(client); err != nil {
+		return nil, err
+	}
+	return client, err
+}
+
+// serviceLogs returns service logs client.
+func (s *CoreClientSafe) serviceLogs(ctx context.Context, in *ServiceLogsRequest, opts ...grpc.CallOption) (Core_ServiceLogsClient, error) {
+	client, err := s.CoreClient.ServiceLogs(ctx, in, opts...)
+	if err != nil {
+		return nil, err
+	}
+	if err := acknowledgement.WaitForStreamToBeReady(client); err != nil {
+		return nil, err
+	}
+	return client, err
+}
+
+// core_ListenEventClient is a client with reconnection.
+type core_ListenEventClient struct {
+	Core_ListenEventClient
+
+	client *CoreClientSafe
+	c      chan *core_ListenEventClientResponse
+
+	ctx  context.Context
+	in   *ListenEventRequest
+	opts []grpc.CallOption
+}
+
+// core_ListenEventClientResponse wraps ListenEvent recv response.
+type core_ListenEventClientResponse struct {
+	eventData *EventData
+	err       error
+}
+
+// newCore_ListenEventClient creates core ListenEvent client.
+func newCore_ListenEventClient(client *CoreClientSafe, ctx context.Context, in *ListenEventRequest, opts ...grpc.CallOption) *core_ListenEventClient {
+	c := &core_ListenEventClient{
+		client: client,
+		c:      make(chan *core_ListenEventClientResponse),
+
+		ctx:  ctx,
+		in:   in,
+		opts: opts,
+	}
+	go c.recvLoop()
+	return c
+}
+
+// recvLoop recives ListenEvent response in loop and reconnect in on error.
+func (s *core_ListenEventClient) recvLoop() {
```

Comment:
Is there any way to not have this function duplicated over and over?

Maybe with a generic version (using `interface{}`) and specific ones that cast the generic?

Reply:
I think this might be over complicated, for generic version we need to provide diffrent:

- api call (`ListenEvent`)
- api call request (`ListenEventRequest`)
- api call response (`ListenEventResponse`)
- data wrapper (`core_ListenEventClientResponse`)
- client (`Core_ListenEventClient`)

So there is a log of different values for 40 line function. I do not like copies of this func as well, but generic is way more complicated there. The best solution will be autogenerate code from protobuf :D